ovs-dpdk.md

```diff
@@ -0,0 +1,170 @@
+To better scale the workloads across cores, multiple pmd threads can be created and pinned to CPU cores by explicitly specifying pmd-cpu-mask. Ensure that cores from the same NUMA nodes are enabled.
+eg: To spawn 2 pmd threads and pin them to cores 1, 2:
+
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:pmd-cpu-mask=6
+
+```
+
+* Add the following entry to  /etc/sysconfig/openvswitch:
+```
+"DPDK_OPTIONS='-c 110 -n 2 --socket-mem 1024,0”
+```
+The pmd-cpu-mask and -c coremask have to be both specified, with the same value.
+ 
+ Restart openvswitch service
+```
+# sudo systemctl restart openvswitch
+```
+ 
+Verify DPDK is enabled:
+```
+# ovs-vsctl get Open_vSwitch . iface_types
+[dpdk, dpdkr, dpdkvhostuser, dpdkvhostuserclient, geneve, gre, internal, ipsec_gre, lisp, patch, stt, system, tap, vxlan]
+```
+
+## OVS-DPDK in oVirt
+ 
+* Upon creating a new cluster, choose OVS as the switch type.
+* When setting up host networks, DPDK-bound NICs are shown in the dialog,as well as Linux-controlled ones
+* When creating a network on DPDK device, an OVS bridge with datapath_type=’netdev’ will be automatically attached to dpdk device will be created.
+
+[ovs-vsctl]: (/images/ovs-dpdk.png)
+ 
+### VMs
+
+VMs will be created with vhostuser network interface. In order to support vhostuser interfaces, hugepages should be configured on the VM.
+ 
+In oVirt GUI, configure hugepages size, number of pages to use and enable shared hugepages.
+Choose ``` hugepages=1048576``` and ```hugepages_shared=true```.
+ 
+ 
+Vhostuser XML file:
+
+``` 
+<interface type="vhostuser">
+            <address bus="0x00" domain="0x0000" function="0x0" slot="0x04" type="pci" />
+            <mac address="00:1a:4a:16:01:54" />
+            <model type="virtio" />
+            <source mode="server" path="/var/run/vdsm/vhostuser/ee844c42-7075-3bf5-86c2-ec831f287c22" type="unix" />
+            <filterref filter="vdsm-no-mac-spoofing" />
+            <link state="up" />
+            <bandwidth />
+        </interface>
+```
+### DPDK in the guest
+
+DPDK can be used in the guest in order to run a high speed packet forwarding program between vhostuser ports.
+VM should be configured to support DPDK using the same steps introduced in the host: install dpdk package, load the suitable kernel module and bind network devices to DPDK.
+Testpmd is a good example of a DPDK application designated for testing packet forwarding.
+
+```# sudo testpmd -l 0-3 --socket-mem 1024 -n 4 -- -i ```
+
+* ```-l``` : List of cores to run on
+* ```--socket-memory```: memory to allocate on specific sockets
+* ```-n``` : number of memory channels to use
+* ```-i``` : interactive mode
+
+For more options, please consult testpmd [documentation](http://dpdk.org/doc/guides/testpmd_app_ug/run_app.html).
+
+
+## Limitations
+* Shared huge pages are required for enabling network communication between VMs and host. They should be configured for every VM using custom properties.
```

Comment:
let's break the first bullet to two, such as:

* Shared huge pages must be configured via the command line for each numa node on each host
* Shared huge pages must be requested for every VM via a custom property

(only now have I understood the annoying per-host complexity. It deserves its own bullet point)